experimental/nomad_resource_plugin/licensedevice/plugin.go

```diff
@@ -40,3 +90,83 @@ func (p *Plugin) Reserve(deviceIDs []string) (*device.ContainerReservation, erro
 func (p *Plugin) Stats(ctx context.Context, interval time.Duration) (<-chan *device.StatsResponse, error) {
 	return nil, fmt.Errorf("Stats not implemented")
 }
+
+func (p *Plugin) fingerprintLoop(ctx context.Context, notifyChan <-chan struct{}, resChan chan<- *device.FingerprintResponse) {
+	for {
+		select {
+		case <-ctx.Done():
+			for {
+				resChan <- &device.FingerprintResponse{Error: fmt.Errorf("context canceled")}
+			}
+			return
+		case _, ok := <-notifyChan:
+			if !ok {
+				return
+			}
+		}
+
+		rctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+		licenses, err := p.notifier.GetCurrent(rctx)
+		cancel()
+		if err != nil {
+			// TODO: metric, log
+		}
+
+		groups, err := deviceGroupsFromLicenses(licenses)
+		if err != nil {
+			// TODO: metric, log?
+		}
+
+		resChan <- &device.FingerprintResponse{Devices: groups}
+	}
+}
+
+func deviceGroupsFromLicenses(ls []*types.License) ([]*device.DeviceGroup, error) {
+	deviceGroupMap := map[string]*device.DeviceGroup{}
+
+	for _, l := range ls {
+		groupName := fmt.Sprintf("%s::%s", l.Vendor, l.Feature)
```

Comment:
`continue` (log to follow) if Vendor or Feature are nil?

Reply:
...they really shouldn't be nil (not null constraints in the DB, etc.) We should ensure that's the case, but I'm not sure this is the best place for such a check

Reply:
True, how do we handle it here? I know Matt adds "required" to the proto files on fields we *need* to exist. But I didn't see that here, nor a check.